datafusion/expr/src/logical_plan/ddl.rs

```diff
@@ -288,8 +290,234 @@ impl PartialOrd for CreateExternalTable {
     }
 }
 
+impl CreateExternalTable {
+    pub fn new(fields: CreateExternalTableFields) -> Result<Self> {
```

Comment:
Given all the fields in `CreateExternalTable` are `pub` anyway, I don't think introducing `CreateExternalTableFields`  requires all construction to go through `new` 🤔 

Therefore I would personally suggest removing the `CreateExternalTableFields` (and equivalent for MemoryTable) and staying with the struct and the builder.

Reply:
> Given all the fields in `CreateExternalTable` are `pub` anyway, I don't think introducing `CreateExternalTableFields` requires all construction to go through `new` 🤔

i think it's disallowed by non_exhaustive.
in fact, i did revert this PR changes in `datafusion/core/src/catalog_common/listing_schema.rs` and `cargo check` failed with

```
error[E0639]: cannot create non-exhaustive struct using struct expression
   --> datafusion/core/src/catalog_common/listing_schema.rs:134:26
    |
134 |                           &CreateExternalTable {
    |  __________________________^
135 | |                             schema: Arc::new(DFSchema::empty()),
136 | |                             name,
137 | |                             location: table_url,
...   |
147 | |                             column_defaults: Default::default(),
148 | |                         },
    | |_________________________^

For more information about this error, try `rustc --explain E0639`.
```

Reply:
Ah, I see -- that was non obvious.

Given that what do you think then about making all the fields not `pub` instead? I think that would be less confusing / more obvious.


Also I do think it is worth considering if we could make a smaller / less breaking change here -- I understand the desire to ensure only semantically correct `CreateExternalTable` structures can be created, but in my opinion the extra cognative overhead of the `*Field` structures makes it significantly harder to work with this code.

What if we
1. Kept the CreateExternalTableBuilder  (to generate errors during construction)
2. Added another  check somewhere when it was processed -- for example in https://github.com/apache/datafusion/blob/eaf51bab8c97341f10f6a46f2ced366a61089d7d/datafusion/core/src/datasource/listing_table_factory.rs#L54 that there were no duplicate columns in the schema